dm: Add marker option to redact info log

db-will opened the pull request:
<!--
Thank you for contributing to TiFlow! 
Please read MD's [CONTRIBUTING](https://github.com/pingcap/tiflow/blob/master/CONTRIBUTING.md) document **BEFORE** filing this PR.
-->

### What problem does this PR solve?
<!--
Please create an issue first to describe the problem.

There MUST be one line starting with "Issue Number:  " and 
linking the relevant issues via the "close" or "ref".

For more info, check https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/contribute-code.html#referring-to-an-issue.
 -->

Issue Number: close #11489

### What is changed and how it works?

Add marker option similar to other components(https://docs.pingcap.com/tidb/stable/log-redaction) for redacting info log

### Check List <!--REMOVE the items that are not applicable-->

#### Tests <!-- At least one of them must be included. -->

Manual test with following config to start dm worker:
```
...
# Log configuration.
log-level = "debug"
log-file = ".log/dm-worker.log"
redact-info-log = "marker"
...
```

By testing with a dm task, we can see the related logs:
```
[2025/02/23 02:04:21.316 -05:00] [DEBUG] [baseconn.go:184] ["execute statement"] [task=testdm] [unit="binlog replication"] [query="INSERT INTO `testdm`.`t1` (`id`,`uid`,`name`,`info`) VALUES (?,?,?,?)"] [argument="‹[20250223 3 Williama <nil>]›"]
```

We are expected to see sensitive data is wrapped by '‹' and '›' marker like other components.


#### Questions <!-- Authors should answer these questions and reviewers should consider these questions. -->

##### Will it cause performance regression or break compatibility?
No

##### Do you need to update user documentation, design documentation or monitoring documentation?
Yes, here are a few documentation needed to be updated:
- https://docs.pingcap.com/tidb/stable/log-redaction
- https://docs.pingcap.com/tidb/stable/dm-worker-configuration-file

### Release note <!-- bugfixes or new features need a release note -->

```release-note
Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

If you don't think this PR needs a release note then fill it with `None`.
```
DM enhances log redaction and supports marking user data in DM worker logs with `‹` and `›`. Based on the marked logs, you can decide whether to redact the marked information when displaying the logs, thus increasing the flexibility of log redaction.

db-will committed:
Add marker option to redact info log

db-will committed:
Fix lint error

lance6716 commented:
/ok-to-test

lance6716 commented on dm/worker/config.go:
```diff
@@ -95,7 +95,8 @@ type Config struct {
 	// - true: enable redact log, which will replace the sensitive information with "?".
-	RedactInfoLog bool `toml:"redact-info-log" json:"redact-info-log"`
+	// - marker: enable redact log, which will use single guillemets ‹› to enclose the sensitive information.
+	RedactInfoLog log.RedactInfoLogType `toml:"redact-info-log" json:"redact-info-log"`
```
I see you have added `UnmarshalTOML` for this new type to ensure compatibility. However I think it's better to add a unit test to prove it.

db-will replied:
i just added extra tests for it

lance6716 reviewed (commented):
Can we reuse TiDB's, or other packages redact functions?

/cc @GMHDBJD @OliverS929

db-will commented on dm/pkg/log/log.go:
```diff
@@ -228,31 +229,158 @@ func WithCtx(ctx context.Context) Logger {
+
+// Type implements pflag.Value interface.
+func (t *RedactInfoLogType) Type() string {
+	return "RedactInfoLogType"
+}
+
+// MarshalJSON implements the `json.Marshaler` interface to ensure the compatibility.
+func (t RedactInfoLogType) MarshalJSON() ([]byte, error) {
+	switch t {
+	case RedactInfoLogON:
+		return json.Marshal(true)
+	case RedactInfoLogMarker:
+		return json.Marshal("MARKER")
+	default:
+	}
+	return json.Marshal(false)
+}
+
+const invalidRedactInfoLogTypeErrMsg = `the "redact-info-log" value is invalid; it should be either false, true, or "MARKER"`
+
+// UnmarshalJSON implements the `json.Marshaler` interface to ensure the compatibility.
+func (t *RedactInfoLogType) UnmarshalJSON(data []byte) error {
+	var s string
+	err := json.Unmarshal(data, &s)
+	if err == nil && strings.ToUpper(s) == "MARKER" {
+		*t = RedactInfoLogMarker
+		return nil
+	}
+	var b bool
+	err = json.Unmarshal(data, &b)
+	if err != nil {
+		return errors.New(invalidRedactInfoLogTypeErrMsg)
+	}
+	if b {
+		*t = RedactInfoLogON
+	} else {
+		*t = RedactInfoLogOFF
+	}
+	return nil
+}
```
implemented

db-will commented on dm/pkg/log/log.go:
```diff
@@ -228,31 +229,158 @@ func WithCtx(ctx context.Context) Logger {
+func (t *RedactInfoLogType) UnmarshalJSON(data []byte) error {
+	var s string
+	err := json.Unmarshal(data, &s)
+	if err == nil && strings.ToUpper(s) == "MARKER" {
+		*t = RedactInfoLogMarker
+		return nil
+	}
+	var b bool
+	err = json.Unmarshal(data, &b)
+	if err != nil {
+		return errors.New(invalidRedactInfoLogTypeErrMsg)
+	}
+	if b {
+		*t = RedactInfoLogON
+	} else {
+		*t = RedactInfoLogOFF
+	}
+	return nil
+}
+
+// UnmarshalTOML implements the `toml.Unmarshaler` interface to ensure the compatibility.
+func (t *RedactInfoLogType) UnmarshalTOML(data any) error {
+	switch v := data.(type) {
+	case bool:
+		if v {
+			*t = RedactInfoLogON
+		} else {
+			*t = RedactInfoLogOFF
+		}
+		return nil
+	case string:
+		if strings.ToUpper(v) == "MARKER" {
+			*t = RedactInfoLogMarker
+			return nil
+		}
+		return errors.New(invalidRedactInfoLogTypeErrMsg)
+	default:
+	}
+	return errors.New(invalidRedactInfoLogTypeErrMsg)
+}
```
implemented

db-will commented:
@OliverS929 could you please take a look at this PR when you have a chance? It has been open for quite a while, and I’ve addressed the previous feedback and added the requested tests. I’d appreciate your review so we can wrap it up.
